torch/distributed/_checkpoint/state_dict_saver.py

```diff
@@ -0,0 +1,213 @@
+        storage_metadata=[StorageMetadata(
+            shard_metadata=shard_metadata,
+            storage_key=fqn,
+            length=storage_size,
+            offset=0,
+        )]
+    )
+    metadata.state_dict_metadata[fqn] = etmd
+
+def _populate_inplace_with_a_sharded_tensor(
+    fqn: str,
+    tensor: ShardedTensor,
+    metadata: Metadata,
+    size_for_storage_handles: Dict[str, int],
+    write_requests: List[ReadWriteRequest],
+):
+    smd = []
+    for shard in tensor.local_shards():
+        # each shard is in it own file.
+        # Most network file system is optimized with single write, multiple read
+        # Unless we can group tensors locally into one big chunk
+        # It might be best to write each shard as one key
+        suffix = "_".join([str(i) for i in shard.metadata.shard_offsets + shard.metadata.shard_sizes])
+        storage_key = f"{fqn}_{suffix}"
+
+        # The reshaping is not required here
+        target_tensor = shard.tensor.detach().reshape(-1,)
+        storage_size = target_tensor.nelement() * target_tensor.element_size()
+
+        one_smd = StorageMetadata(
+            shard_metadata=shard.metadata,
+            storage_key=storage_key,
+            length=storage_size,
+            offset=0,
+        )
+        smd.append(one_smd)
+
+        size_for_storage_handles[storage_key] = storage_size
+
+        wr = ReadWriteRequest(
+            target_tensor=target_tensor,
+            storage_key=storage_key,
+            offset=0,
+            length=storage_size,
+        )
+        write_requests.append(wr)
+
+
+    etmd = ExtendedTensorMetadata(
+        tensor_metadata=tensor.metadata(),
+        storage_metadata=smd,
+    )
+    metadata.state_dict_metadata[fqn] = etmd
+
+
+def _prepare(state_dict: Dict[str, Any]) -> Tuple[Metadata, Dict[str, int], List[ReadWriteRequest]]:
+    """
+    Uses the state_dict to build three things.
+
+    metadata: Metadata
+        The metatdata discribing the tensor / sharded tensor.
+        And it is storage meta data. See "../metadata.py" for detail
+
+    size_for_storage_handles: Dict[str, int]
+        Key is the handle name, value is the handle's size
+        It can used to pre allocate the storage for parallel and non sequential writes.
+
+    write_requests: List[ReadWriteRequest]
+        List of write requests that should br perfromed by the writer.
+
+
+    Subclasses can optionally overwrite the implementation here,
+    if the default does not meet its requirement.
+    """
+    metadata = Metadata(state_dict_metadata={})
+    storage_handles = {}
+    write_requests = []
+
+    for fqn, tensor in state_dict.items():
+        if isinstance(tensor, Tensor):
+            # The assumption is that non ShardedTensors are full replicated across all ranks
+            # So we just need one from Rank 0.
+            # If that's not the case, we will update later.
+            if dist.is_initialized() and dist.get_rank() != 0:
```

Comment:
Shouldn't we try to minimize the load across all ranks?
Something like `hash(fqn) % dist.world_size() != dist.get_world_size()` instead?

Reply:
Of course, I am just doing whatever that is simplest.  There will be some implication on the call sequence w.r.t. checkpoint load and  DDP, alltoall vs broadcast. This will require more discussion.